cli/src/commands/deploy.js

```diff
@@ -0,0 +1,186 @@
+const { reporter, chalk } = require('@dhis2/cli-helpers-engine')
+const path = require('path')
+const fs = require('fs-extra')
+const FormData = require('form-data')
+const inquirer = require('inquirer')
+
+const makePaths = require('../lib/paths')
+const parseConfig = require('../lib/parseConfig')
+const { createClient } = require('../lib/httpClient')
+
+const constructAppUrl = (baseUrl, config, serverVersion) =>
+    baseUrl +
+    (config.coreApp ? '/dhis-web-' : '/api/apps/') +
+    (serverVersion.minor < 35 ? config.title : config.name)
+        .replace(/[^A-Za-z0-9\s-]/g, '')
+        .replace(/\s+/g, '-') +
+    (serverVersion.minor < 35 ? '/index.html' : '/')
+
+const dumpHttpError = (message, response) => {
+    if (!response) {
+        reporter.error(message)
+        return
+    }
+
+    reporter.error(
+        message,
+        response.status,
+        typeof response.data === 'object'
+            ? response.data.message
+            : response.statusText
+    )
+    reporter.debugErr('Error details', response.data)
+}
+
+const promptForInputs = async params => {
+    if (
+        process.env.CI &&
+        (!params.baseUrl || !params.username || !params.password)
+    ) {
+        reporter.error(
+            'Prompt disabled in CI mode - missing baseUrl, username, or password parameter.'
+        )
+        process.exit(1)
+    }
+    return await inquirer.prompt([
+        {
+            type: 'input',
+            name: 'baseUrl',
+            message: 'DHIS2 instance URL:',
+            when: () => !params.baseUrl,
+            validate: input =>
+                !input ||
+                !input.length ||
+                (!input.startsWith('http://') && !input.startsWith('https://'))
+                    ? 'Please enter a valid URL'
+                    : true,
```

Comment:
I'd find the reverse a lot easier to digest:
```javascript
            validate: input =>
                input &&
                input.length &&
                (input.startsWith('http://') || input.startsWith('https://'))
                    ? true
                    : 'Please enter a valid URL',
```

Reply:
Done, factored into helper function